datalad/core/distributed/push.py

```diff
@@ -0,0 +1,726 @@
+        jobs=jobs_opt,
+    )
+
+    # Desired features:
+    # - let Git do its thing (push multiple configured refs without the need
+    #                         to specific anything on the command line)
+    #   - compilication: we need publication dependencies (i.e. publish what
+    #     would be published by Git to a different remote first, hence we
+    #     cannot simply watch Git do it, and later act on it.)
+    #   - https://github.com/datalad/datalad/issues/1284
+    #   - https://github.com/datalad/datalad/issues/4006
+    # - make differences between remotes and various types of special remotes
+    #   opaque
+    #   - https://github.com/datalad/datalad/issues/3127
+    # - informative and comprehensive (error) reporting
+    #   - https://github.com/datalad/datalad/issues/2000
+    #   - https://github.com/datalad/datalad/issues/1682
+    #   - https://github.com/datalad/datalad/issues/2029
+    #   - https://github.com/datalad/datalad/issues/2855
+    #   - https://github.com/datalad/datalad/issues/3412
+    #   - https://github.com/datalad/datalad/issues/3424
+    # - ensure robust behavior in multi-lateral push scenarios (updating
+    #   a dataset that was updated by a 3rd-party after the last known
+    #   fetched change
+    #   - https://github.com/datalad/datalad/issues/2636
+    # - should NOT mimic `publish` in that it mixes `create-sibling` and
+    #   `push` into a single operation. This would fold the complexity
+    #   of all possible ways a local dataset hierarchy could possibly
+    #   connect to remote ends into this command. It would be a lost battle
+    #   from the start.
+    #   - not tackle: https://github.com/datalad/datalad/issues/2186
+    # - maintain standard setup, and not reflect procedural aspects
+    #   onto the resulting outcomes
+    #   - https://github.com/datalad/datalad/issues/2001
+    # - do a straight push, nothing like 'sync'. If a remote has something that
+    #   needs merging first, fail and let users update. Any diff we are missing
+    #   locally can impact decision making via --since and friends.
+
+    @staticmethod
+    @datasetmethod(name='push')
+    @eval_results
+    def __call__(
+            path=None,
+            dataset=None,
+            to=None,
+            since=None,
+            force=None,
+            recursive=False,
+            recursion_limit=None,
+            jobs=None):
+        # we resolve here, because we need to perform inspection on what was given
+        # as an input argument further down
+        paths = [resolve_path(p, dataset) for p in ensure_list(path)]
+
+        ds = require_dataset(
+            dataset, check_installed=True, purpose='pushing')
+
+        if since:
+            # will blow with ValueError if unusable
+            ds.repo.get_hexsha(since)
+
+        if not since and since is not None:
+            # TODO figure out state of remote branch and set `since`
+            pass
```

Comment:
If we're going to merge this in as is, should the mention of the not-yet-implemented empty string behavior be removed from `since`s description?

Reply:
I have imported the behavior of `publish`, so the comment is valid now.